### ISSUE TITLE
Layer-wise LR decay implementation

### ISSUE THREAD

**Issue by umbertocappellazzo** [SihanXU/nepa#11]
Hello! Thanks for the great paper and releasing the code. 

I have one question regarding the layer-wise LR decay implementation used during fine-tuning. In the paper, you say that you implement a layer-wise LR decay scheduler, which linearly increases the decay from a small initial value (0.35 based on Table 8) to 1.0 over the course of fine-tuning. You compare this variant against the configuration with a decay factor equal to 0.65 kept constant throughout. The results with the layer-wise decay is 83.8 compared to 83.0 without it. 83.8 is also the value reported in the main table 2.

I think in the current implementation, by default you are using the configuration with constant decay factor == 0.65 since in `nepa_b_sft_sh` I see `--llrd 0.65` and `--lr_scheduler_kwargs '{"custom_scheduler_type": "llrd_cosine_warmup"}`. By inspecting the `llrd_cosine_warmup` implementation (`get_llrd_cosine_schedule_with_warmup` in the `schedulers.py` script), the parameter `llrd_end` is not defined, whereas in `get_llrd_cosine_schedule` the parameter is defined and set to None as default. I think that if we wanted to apply the layer-wise LR decay we should define `--llrd_end` and set it to 1.0 and update the `get_llrd_cosine_schedule_with_warmup` accordingly? Is that correct? Since in the paper you mention multiple times the layer-wise LR decay scheduler is used, I think by default the code should rely on the layer-wise LR decay. Thank you!

Umberto

**Comment by SihanXU** [comment]
Thank you for the careful reading and for pointing this out and sorry for the confusion.

The main clarification is that the released repo is a re-implemented version, not the exact training code used for the paper. In particular, the released implementation does not use SwiGLU. In our original experiments, the stronger overfitting with SwiGLU made the layer-wise LR decay scheduler especially helpful, which is why it is emphasized in the paper.

During the re-implementation, we found that with GeLU, the original fixed LLRD setting was already sufficient to reach essentially the same performance, so the default config in the repo keeps --llrd 0.65 and does not use the progressive decay schedule by default.

**Comment by umbertocappellazzo** [comment]
Thank you for your clarification. Actually, I just saw that this detail is highlighted in the README.md, I glossed over it, my bad. 

Is this the only difference with respect to the original paper implementation?

**Comment by SihanXU** [comment]
I think there is one more change in the fine-tuning stage: I added pooling layer (https://github.com/SihanXU/nepa/blob/main/configs/finetune/nepa-base-patch14-224-sft/config.json#L2, https://github.com/SihanXU/nepa/blob/main/models/vit_nepa/modeling_vit_nepa.py#L931) in the classification head output because the other models (MAE, BEiT) follow the same pattern.

**Comment by umbertocappellazzo** [comment]
Oh I see! By inspecting your script (https://github.com/SihanXU/nepa/blob/main/models/vit_nepa/modeling_vit_nepa.py#L895), I notice you define `self.pooler` but then at line 933 you explicitly use `image_tokens.mean(dim=1)`. Also, at line 936, you write `pooled_output = sequence_output[:, -1, :]` but I guess it should be `[:, 0, :]` if you want to use the CLS token? Maybe a typo since I think you always use the pooling version.

**Comment by SihanXU** [comment]
The use of sequence_output[:, -1, :] is for compatibility with the causal masking setting. Under causal masking, only the last embedding has access to the full preceding context, so it is the one that contains the complete information. In that case, the CLS token is not used as a true classification token.
Here, the CLS token plays a role closer to a BOS token in an LLM, rather than a standard classification token that should be read out at the end. So using [:, 0, :] would not be appropriate for the causal masking case.

**Comment by umbertocappellazzo** [comment]
I see, it makes sense. You also mention this in the second paragraph of section 3.3. I was misled by the fact that during fine-tuning you use bidirectional attention, so it is not causal. I think if you use bidirectional attention, you can use both the last token or the mean of all image tokens, with the latter which should yield better performance. If you use causal masking during fine-tuning as you did in the ablation study, then it makes sense to only use the last token. 
Either way, did you notice better results with the pooling layer than the last token, right?

**Comment by umbertocappellazzo** [comment]
I'm closing the issue as I'm satisfied with the clarifications. Thank you again!
